zerver/tests/test_email_change.py

```diff
@@ -34,6 +34,17 @@ def test_confirm_email_change_with_non_existent_key(self):
         self.assertEqual(response.status_code, 200)
         self.assertIn("Whoops", response.content.decode('utf8'))
 
+    def test_email_change_without_logging_in(self):
+        # type: () -> None
+        key = generate_key()
+        with self.assertRaises(EmailChangeConfirmation.DoesNotExist):
+            url = EmailChangeConfirmation.objects.get_activation_url(key)
+
+        url = EmailChangeConfirmation.objects.get_activation_url(
+            key, 'testserver')
+        response = self.client_get(url)
+        self.assertEqual(response.status_code, 302)
```

Comment:
The test you've written test is great.  You need two more tests to confirm the new feature actually works.

You should add another test that is similar to `test_end_to_end_flow`, but checks that it fails when the user tries to do the `/json/settings/change` request, and verifies nothing got put in the outbox if the new setting is used.

And a second test that has the user go through the first couple steps (through getting the email), and then change the realm-level setting, and confirm the email confirmation link doesn't work.

I think actually that first test will find a bug in your code -- you've just disabled the UI on the frontend, but someone sneaky could use the endpoint (powered by the `/json/settings/change` view) to start the flow to change their email anyway, which would be confusing.  Will need to add a check and error message in `json_change_settings`.